bot/utils/helpers.py

```diff
@@ -0,0 +1,71 @@
+import asyncio
+from typing import Any, Callable, Iterable, Optional, TypeVar
+
+import discord
+import more_itertools
+
+from bot import logger
+
+T = TypeVar('T')
+
+
+async def chunked_find(
+    predicate: Callable[[T], Any],
+    seq: Iterable[T],
+    *,
+    chunk_size: Optional[int] = None
+) -> Optional[T]:
+    """
+    A helper to return the first element found in the sequence that meets the predicate.
+
+    If chunk_size is specified, chunked_find() will yield control to the event loop between each chunk.
+    """
+    for chunk in more_itertools.chunked(seq, chunk_size):
+        for element in chunk:
+            if predicate(element):
+                return element
+
+        if chunk_size:
+            await asyncio.sleep(0)  # Yield to the event loop
+    return None
+
+
+async def _check_first_message_referencing(thread: discord.Thread, message_id: int) -> Optional[discord.Thread]:
+    """Returns a thread if the thread_starter_message in `thread` references the message_id."""
+    messages = await thread.history(limit=1, oldest_first=True).flatten()
+    if messages[0].reference.message_id == message_id:
+        return thread
+    return None
+
+
+async def get_thread_from_message_id(
+    message_id: int,
+    channel: discord.TextChannel,
+    cached_messages: list[discord.Message]
+) -> Optional[discord.Thread]:
+    """Returns the thread linked to the given message id."""
+    def predicate(message: discord.Message) -> bool:
+        """Returns true if `message` is the thread_starter_message for the `message_id`."""
+        return (
+            message.reference
+            and message.reference.message_id == message_id
+            and isinstance(message.channel, discord.Thread)
+        )
+
+    # Try and find the thread start messag ein message cache.
+    # Use a chunked find as this many checks could block for too long.
+    if message := await chunked_find(predicate, cached_messages, chunk_size=200):
+        logger.info("Thread found in message cache!")
+        return message.channel
+
+    for thread in channel.threads:  # Thread may be in cache
+        if found := await _check_first_message_referencing(thread, message_id):
+            logger.info("Thread found in thread cache!")
+            return found
+
+    logger.info("Message not found in either cache, fetching all threads...")
+    for thread in await channel.active_threads():
+        if found := await _check_first_message_referencing(thread, message_id):
+            logger.info("Thread found in fetched threads!")
+            return found
+    return None
```

Comment:
I would like a blank like to kind of seperate this.
```suggestion
            return found

    return None
```